File: DFS/46permutations.py
class Solution(object):
    def permute(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        def dfs(nums, path, res):
            if not nums:
                res.append(path)
            for i in range(len(nums)):
                print(nums[:i]+nums[i+1:])
                dfs(nums[:i]+nums[i+1:], path+[nums[i]],res)
        
        res = []
        dfs(nums,[],res)
        return 0

File: DFS/46Permutations.cpp
class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> res;
        vector<int> temp;
        vector<bool> used(nums.size(),false);
        helper(res, temp, nums, used);
        return res;
    }
    
    void helper(vector<vector<int>>& res, vector<int>& temp, vector<int>& nums, vector<bool>& used){
        if(temp.size()==nums.size()){
            res.push_back(temp);
        } 
        else{
            for(int i=0; i<nums.size(); i++){
                if(!used[i]){
                    temp.push_back(nums[i]);
                    used[i]=true;
                    helper(res, temp, nums, used);
                    used[i]=false;
                    temp.pop_back();
                }   
            }
        }
    }
};